File: sdk/unity/com.openmasu.sdk/Runtime/Plugins/iOS/Sources/OpenMasuMax/MaxRevenueAdapter.swift
import Foundation
#if canImport(OpenMasuCore)
import OpenMasuCore
#endif

public struct MaxRevenueObservation: Sendable {
  public let revenue: Double
  public let precision: String
  public let networkName: String
  public let adUnitId: String
  public let format: String
  public let placement: String?
  public let networkPlacement: String?

  public init(
    revenue: Double,
    precision: String,
    networkName: String,
    adUnitId: String,
    format: String,
    placement: String? = nil,
    networkPlacement: String? = nil
  ) {
    self.revenue = revenue
    self.precision = precision
    self.networkName = networkName
    self.adUnitId = adUnitId
    self.format = format
    self.placement = placement
    self.networkPlacement = networkPlacement
  }
}

public final class MaxRevenueMapper: @unchecked Sendable {
  private let installationId: @Sendable () throws -> String
  private let impressionId: @Sendable () -> String
  private let lock = NSLock()
  private var rejected = 0

  public init(
    installationId: @escaping @Sendable () throws -> String,
    impressionId: (@Sendable () -> String)? = nil
  ) {
    self.installationId = installationId
    self.impressionId = impressionId ?? { "impression:\(UuidV7.generate().uuidString.lowercased())" }
  }

  public var errorCount: Int {
    lock.lock()
    defer { lock.unlock() }
    return rejected
  }

  public func map(_ observation: MaxRevenueObservation) throws -> [String: Any]? {
    guard observation.revenue.isFinite, observation.revenue >= 0,
          ["publisher_defined", "exact", "estimated", "undefined"].contains(observation.precision),
          !observation.networkName.isEmpty, !observation.adUnitId.isEmpty
    else {
      lock.lock(); rejected += 1; lock.unlock()
      return nil
    }
    guard let decimalRevenue = Decimal(string: String(observation.revenue), locale: Locale(identifier: "en_US_POSIX")) else {
      lock.lock(); rejected += 1; lock.unlock()
      return nil
    }
    var value = decimalRevenue * Decimal(1_000_000)
    var rounded = Decimal()
    NSDecimalRound(&rounded, &value, 0, .bankers)
    var extensions: [String: Any] = ["ad_format": observation.format]
    if let placement = observation.placement { extensions["placement"] = placement }
    if let networkPlacement = observation.networkPlacement { extensions["network_placement"] = networkPlacement }
    return [
      "event_name": "ad_revenue",
      "subject_scope": "installation_level",
      "installation_id": try installationId(),
      "impression_id": impressionId(),
      "ad_unit_id": observation.adUnitId,
      "ad_network": observation.networkName,
      "mediation_provider": "applovin-max",
      "amount_unscaled": NSDecimalNumber(decimal: rounded).stringValue,
      "amount_scale": 6,
      "currency": "USD",
      "currency_source": "reported",
      "revenue_source": "client_estimated",
      "revenue_precision": observation.precision,
      "extensions": extensions,
    ]
  }
}

public struct OpenMasuMaxAdapter: Sendable {
  private let sdk: OpenMasuSDK
  private let mapper: MaxRevenueMapper

  public init(sdk: OpenMasuSDK, mapper: MaxRevenueMapper) {
    self.sdk = sdk
    self.mapper = mapper
  }

  @discardableResult
  public func didPayRevenue(_ observation: MaxRevenueObservation) async throws -> Bool {
    guard let payload = try mapper.map(observation) else { return false }
    guard let eventId = payload["impression_id"] as? String else { return false }
    try await sdk.enqueueAdRevenue(payload: payload, eventId: eventId)
    return true
  }
}

private enum UuidV7 {
  static func generate(now: Date = Date()) -> UUID {
    let milliseconds = UInt64(max(0, now.timeIntervalSince1970 * 1_000))
    var bytes = (0..<16).map { _ in UInt8.random(in: 0...255) }
    for index in 0..<6 { bytes[index] = UInt8((milliseconds >> UInt64((5 - index) * 8)) & 0xff) }
    bytes[6] = (bytes[6] & 0x0f) | 0x70
    bytes[8] = (bytes[8] & 0x3f) | 0x80
    return UUID(uuid: (
      bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6], bytes[7],
      bytes[8], bytes[9], bytes[10], bytes[11], bytes[12], bytes[13], bytes[14], bytes[15]
    ))
  }
}

File: sdk/android/max/src/main/java/dev/openmmp/sdk/max/MaxRevenueAdapter.kt
package dev.openmmp.sdk.max

import com.applovin.mediation.MaxAd
import com.applovin.mediation.MaxAdRevenueListener
import dev.openmmp.sdk.OpenMmpSdk
import org.json.JSONObject
import java.math.BigDecimal
import java.math.RoundingMode
import java.security.SecureRandom
import java.util.UUID
import java.util.concurrent.atomic.AtomicLong

data class MaxRevenueObservation(
  val revenue: Double,
  val precision: String,
  val networkName: String,
  val adUnitId: String,
  val placement: String? = null,
  val networkPlacement: String? = null,
  val format: String = "unknown",
)

class MaxRevenueMapper(
  private val installationId: () -> String,
  private val impressionId: () -> String = { "impression:${UuidV7.generate()}" },
) {
  val errorCount = AtomicLong(0)

  fun map(observation: MaxRevenueObservation): JSONObject? {
    if (!observation.revenue.isFinite() || observation.revenue < 0.0) {
      errorCount.incrementAndGet()
      return null
    }
    val precision = observation.precision.takeIf { it in PRECISIONS }
    if (precision == null) {
      errorCount.incrementAndGet()
      return null
    }
    val micros = BigDecimal.valueOf(observation.revenue)
      .movePointRight(6)
      .setScale(0, RoundingMode.HALF_EVEN)
      .toPlainString()
    return JSONObject()
      .put("event_name", "ad_revenue")
      .put("subject_scope", "installation_level")
      .put("installation_id", installationId())
      .put("impression_id", impressionId())
      .put("ad_unit_id", observation.adUnitId)
      .put("ad_network", observation.networkName)
      .put("mediation_provider", "applovin-max")
      .put("amount_unscaled", micros)
      .put("amount_scale", 6)
      .put("currency", "USD")
      .put("currency_source", "reported")
      .put("revenue_source", "client_estimated")
      .put("revenue_precision", precision)
      .put("extensions", JSONObject().apply {
        put("ad_format", observation.format)
        observation.placement?.let { put("placement", it) }
        observation.networkPlacement?.let { put("network_placement", it) }
      })
  }

  private companion object {
    val PRECISIONS = setOf("publisher_defined", "exact", "estimated", "undefined")
  }
}

class OpenMmpMaxRevenueListener(
  private val sdk: OpenMmpSdk,
  private val mapper: MaxRevenueMapper,
) : MaxAdRevenueListener {
  override fun onAdRevenuePaid(ad: MaxAd) {
    mapper.map(
      MaxRevenueObservation(
        revenue = ad.revenue,
        precision = ad.revenuePrecision,
        networkName = ad.networkName,
        adUnitId = ad.adUnitId,
        placement = ad.placement,
        networkPlacement = ad.networkPlacement,
        format = ad.format.label,
      ),
    )?.let { payload -> sdk.enqueueAdRevenue(payload, payload.getString("impression_id")) }
  }
}

object OpenMmpMaxBridge {
  @JvmStatic
  fun listener(sdk: OpenMmpSdk): OpenMmpMaxRevenueListener =
    OpenMmpMaxRevenueListener(sdk, MaxRevenueMapper(sdk::installationIdForMeasurement))

  @JvmStatic
  fun track(
    sdk: OpenMmpSdk,
    revenue: Double,
    precision: String,
    networkName: String,
    adUnitId: String,
    placement: String?,
    networkPlacement: String?,
  ): Boolean {
    val payload = MaxRevenueMapper(sdk::installationIdForMeasurement).map(
      MaxRevenueObservation(revenue, precision, networkName, adUnitId, placement, networkPlacement),
    ) ?: return false
    sdk.enqueueAdRevenue(payload, payload.getString("impression_id"))
    return true
  }
}

private object UuidV7 {
  private val random = SecureRandom()

  fun generate(nowMs: Long = System.currentTimeMillis()): UUID {
    val bytes = ByteArray(16).also(random::nextBytes)
    for (index in 0 until 6) bytes[index] = ((nowMs ushr ((5 - index) * 8)) and 0xff).toByte()
    bytes[6] = ((bytes[6].toInt() and 0x0f) or 0x70).toByte()
    bytes[8] = ((bytes[8].toInt() and 0x3f) or 0x80).toByte()
    var most = 0L
    var least = 0L
    for (index in 0 until 8) most = (most shl 8) or (bytes[index].toLong() and 0xff)
    for (index in 8 until 16) least = (least shl 8) or (bytes[index].toLong() and 0xff)
    return UUID(most, least)
  }
}
